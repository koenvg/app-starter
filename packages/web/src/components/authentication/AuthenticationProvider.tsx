import { login } from "./api";
import React, { SFC } from "react";
import { useLocation, useHistory } from "react-router-dom";

export interface User {
  username: string;
}
interface Authentication {
  isAuthenticated: boolean;
  user?: User;
  authenticate: (username: string, password: string) => Promise<void>;
}

const AuthenticationContext = React.createContext<Authentication>({
  isAuthenticated: false,
  authenticate: () => {
    throw new Error("not implemented");
  },
});

export const AuthenticationProvider: SFC = ({ children }) => {
  const [isAuthenticated, setIsAuthenticated] = React.useState<boolean>(false);
  const [user, setUser] = React.useState<User>();
  const { state } = useLocation();
  const history = useHistory();

  const authenticate = React.useCallback<Authentication["authenticate"]>(
    async (username, password) => {
      const user = await login(username, password);
      setUser(user);
      setIsAuthenticated(true);
      if (state.from) {
        history.push(state.from);
      }
    },
    [history, state],
  );

  return (
    <AuthenticationContext.Provider value={{ isAuthenticated, user, authenticate }}>
      {children}
    </AuthenticationContext.Provider>
  );
};

export const useAuthentication = () => React.useContext(AuthenticationContext);

AuthenticationProvider.displayName = "AuthenticationProvider";
