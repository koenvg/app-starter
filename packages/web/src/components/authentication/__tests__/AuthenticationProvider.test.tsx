import { AuthenticationProvider, useAuthentication } from "../AuthenticationProvider";
import React from "react";
import { render, wait, fireEvent } from "@testing-library/react";
import { Router } from "react-router-dom";
import { createMemoryHistory, MemoryHistoryBuildOptions } from "history";

describe("AuthenticationProvider", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const LoginButton = () => {
    const { authenticate, user, isAuthenticated } = useAuthentication();

    return (
      <>
        <div>User: {JSON.stringify(user)}</div>
        <div>isAuthenticated: {new String(isAuthenticated)}</div>
        <button onClick={() => authenticate("username", "password")} />
      </>
    );
  };

  it("Should log the user in and navigate to the page in the state", async () => {
    const history = createMemoryHistory(({
      initialEntries: [{ pathname: "/login", state: { from: "/" } }],
    } as unknown) as MemoryHistoryBuildOptions);
    const component = render(
      <Router history={history}>
        <AuthenticationProvider>
          <LoginButton />
        </AuthenticationProvider>
      </Router>,
    );
    expect(component.container).toMatchSnapshot();

    await wait(() => {
      fireEvent.click(component.container.querySelector("button") as Element);
    });

    expect(history.location.pathname).toBe("/");

    expect(component.container).toMatchSnapshot();
  });
});
