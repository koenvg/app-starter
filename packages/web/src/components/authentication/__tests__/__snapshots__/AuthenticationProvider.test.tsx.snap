// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthenticationProvider Should log the user in and navigate to the page in the state 1`] = `
<div>
  <div>
    User: 
  </div>
  <div>
    isAuthenticated: 
    f
    a
    l
    s
    e
  </div>
  <button />
</div>
`;

exports[`AuthenticationProvider Should log the user in and navigate to the page in the state 2`] = `
<div>
  <div>
    User: 
    {"username":"username"}
  </div>
  <div>
    isAuthenticated: 
    t
    r
    u
    e
  </div>
  <button />
</div>
`;
